File: backend/src/main/java/com/construhogar/backend/service/InventarioService.java
package com.construhogar.backend.service;

import java.util.List;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.construhogar.backend.model.Inventario;
import com.construhogar.backend.model.Producto;
import com.construhogar.backend.repository.InventarioRepository;
import com.construhogar.backend.repository.ProductoRepository; // Importante para buscar el material

@Service
public class InventarioService {

    private final InventarioRepository inventarioRepository;
    private final ProductoRepository productoRepository; // Inyectamos el repositorio de productos

    public InventarioService(InventarioRepository inventarioRepository, ProductoRepository productoRepository) {
        this.inventarioRepository = inventarioRepository;
        this.productoRepository = productoRepository;
    }

    // ======================
    // CRUD BÁSICO
    // ======================

    public List<Inventario> obtenerTodos() {
        return inventarioRepository.findAll();
    }

    public Inventario obtenerPorId(Integer id) {
        return inventarioRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Inventario no encontrado"));
    }

    public Inventario guardar(Inventario inventario) {
        return inventarioRepository.save(inventario);
    }

    public void eliminar(Integer id) {
        inventarioRepository.deleteById(id);
    }

    // ======================
    // LÓGICA DE NEGOCIO
    // ======================

    /**
     * Este método es el que usará el Frontend para las entradas.
     * Busca el producto por su ID y luego aumenta el stock.
     */
    @Transactional
    public void registrarEntrada(Integer idProducto, Integer cantidad) {
        // 1. Verificamos que el producto exista en la base de datos
        Producto producto = productoRepository.findById(idProducto)
                .orElseThrow(() -> new RuntimeException("El producto con ID " + idProducto + " no existe."));
        
        // 2. Llamamos a la lógica de aumento de stock
        this.aumentarStock(producto, cantidad);
    }

    /**
     * 🔥 NUEVO MÉTODO: Registrar salida de material (Ajuste manual)
     * Busca el producto y reduce el stock validando existencias.
     */
    @Transactional
    public void registrarSalida(Integer idProducto, Integer cantidad) {
        // Verificamos que el producto exista
        if (!productoRepository.existsById(idProducto)) {
            throw new RuntimeException("El producto con ID " + idProducto + " no existe.");
        }
        // Ejecutamos la reducción de stock
        this.reducirStock(idProducto, cantidad);
    }

    // 🔼 COMPRA → aumenta stock (Lógica interna)
    @Transactional
    public void aumentarStock(Producto producto, Integer cantidad) {

        Inventario inventario = inventarioRepository
                .findByProducto(producto)
                .orElseGet(() -> {
                    // Si el producto no tiene registro en inventario, lo creamos desde cero
                    Inventario nuevo = new Inventario();
                    nuevo.setProducto(producto);
                    nuevo.setStockActual(0);
                    return nuevo;
                });

        // Sumamos la nueva cantidad al stock que ya existía
        inventario.setStockActual(
                inventario.getStockActual() + cantidad
        );

        inventarioRepository.save(inventario);
    }

    // 🔽 VENTA / AJUSTE → reduce stock
    @Transactional
    public void reducirStock(Integer idProducto, Integer cantidad) {

        Inventario inventario = inventarioRepository
                .findByProducto_IdProducto(idProducto)
                .orElseThrow(() -> new RuntimeException("Producto sin inventario"));

        if (inventario.getStockActual() < cantidad) {
            throw new RuntimeException("Stock insuficiente en bodega");
        }

        inventario.setStockActual(
                inventario.getStockActual() - cantidad
        );

        inventarioRepository.save(inventario);
    }
}

File: frontend/desing/construhogar-frontend/src/services/inventarioService.js
import axios from 'axios';

// URL base que apunta a tu controlador de Inventario en Java
const API_URL = "http://localhost:8080/api/inventarios";

export const inventarioService = {

  // 1. Obtener todo el inventario (útil para la tabla general)
  listarTodo: async () => {
    const res = await axios.get(API_URL);
    return res.data;
  },

  // 2. Consultar stock detallado por el ID del Material/Producto
  consultarPorProducto: async (idProducto) => {
    const res = await axios.get(`${API_URL}/producto/${idProducto}`);
    return res.data;
  },

  // 3. Registrar la entrada de stock (Sumar cantidades por ajuste o compra)
  registrarEntrada: async (idProducto, cantidad) => {
    const res = await axios.post(`${API_URL}/entrada`, {
      idProducto: parseInt(idProducto),
      cantidad: parseInt(cantidad)
    });
    return res.data;
  },

  // 4. 🔥 NUEVA FUNCIÓN: Registrar la salida de stock (Ajuste manual por error)
  // Envía el idProducto y la cantidad al nuevo endpoint de salida en el Backend
  registrarSalida: async (idProducto, cantidad) => {
    const res = await axios.post(`${API_URL}/salida`, {
      idProducto: parseInt(idProducto),
      cantidad: parseInt(cantidad)
    });
    return res.data;
  },

  // 5. Actualizar datos básicos (como cambiar la ubicación en bodega)
  actualizarInventario: async (idInventario, datos) => {
    const res = await axios.put(`${API_URL}/${idInventario}`, datos);
    return res.data;
  },

  // 6. Eliminar un registro de inventario
  eliminar: async (idInventario) => {
    await axios.delete(`${API_URL}/${idInventario}`);
  }
};